datafusion-cli/src/object_storage.rs

```diff
@@ -43,28 +46,52 @@ impl FromStr for ObjectStoreScheme {
     }
 }
 
-#[derive(Debug)]
-pub struct DatafusionCliObjectStoreProvider {}
+#[derive(Debug, Default)]
+pub struct DatafusionCliObjectStoreRegistry {
+    inner: DefaultObjectStoreRegistry,
+}
+
+impl DatafusionCliObjectStoreRegistry {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl ObjectStoreRegistry for DatafusionCliObjectStoreRegistry {
+    fn register_store(
+        &self,
+        key: &str,
```

Comment:
What is the use-case for registering stores not based on scheme and host? I mainly ask as it is unclear how this method interacts with get_by_url? Could we just keep the host and scheme signature instead?

Reply:
> the key for registration is

That's the key for lookup, and so it was harmless for it to contain more, it would just not be found.

> port

Including an optional port seems like a sensible addition

Reply:
If with only scheme and host for register, how can we achieve register by url which may be invoked in `get_with_url` for lazy registration with key of `&url[url::Position::BeforeScheme..url::Position::BeforePath]`

Reply:
That would be a bug then, yes. We shouldn't be registering with that, well spotted.

It should instead call register_store with just the relevant parts of the URL

Reply:
Alternatively we could just make the register_store take a `Url` as the key, and leave it to the `ObjectStoreRegistry` to define what parts it uses. I don't know, this API has morphed so much from its original design now :sweat_smile:

Reply:
I'm OK with either with url or a pair of (scheme, host, port).

Reply:
Let's go with a URL, it is more flexible, avoids a potentially breaking change even if it was unintentional, and also matches the get method

Reply:
Thank you, apologies for the back and forth :smile:

Reply:
Np. We are making things better and better. Thanks for your suggestions to let the interface simply and meaningful.